### PR TITLE
Remove std components from Debug impls when --use-core is given

### PR DESCRIPTION
**Pull request by chrysn**
This implements a solution to https://github.com/rust-lang-nursery/rust-bindgen/issues/1100. The ::std:: vs ::core:: parts in the trait names are trivial; the format!/String usage is so far only patched out (ie. Debug gives "[...]" rather than actual data on the affected fields).

### CODE REVIEW

File: src/codegen/mod.rs [hunk 1]
```diff
@@ -1918,8 +1918,10 @@ impl CodeGenerator for CompInfo {
                 self.kind(),
             );
 
+            let fmtparent = if ctx.options().use_core { quote! { core } } else { quote! { std } };
```
Comment (emilio) [inline src/codegen/mod.rs]: `ctx.trait_prefix()` instead, and let's call this `prefix` for consistency with other `codegen` code.

File: src/codegen/impl_debug.rs [hunk 2]
```diff
@@ -184,28 +186,40 @@ impl<'a> ImplDebug<'a> for Item {
                     // The simple case
                     debug_print(name, quote! { #name_ident })
                 } else {
-                    // Let's implement our own print function
+                    if ctx.options().use_core {
+                        Some((
+                            format!("{}: [...]", name), vec![]
+                            ))
```
Comment (emilio) [inline src/codegen/impl_debug.rs]: nit: indentation is off.

File: src/codegen/impl_debug.rs [hunk 3]
```diff
@@ -184,28 +186,40 @@ impl<'a> ImplDebug<'a> for Item {
                     // The simple case
                     debug_print(name, quote! { #name_ident })
                 } else {
-                    // Let's implement our own print function
+                    if ctx.options().use_core {
```
Comment (emilio) [inline src/codegen/impl_debug.rs]: Maybe point out that there's no `String` in `core`.

Comment (chrysn) [reply src/codegen/impl_debug.rs]: Done.

File: src/codegen/mod.rs (hunk 1 above)
Comment (chrysn) [reply src/codegen/mod.rs]: Thanks, done; I was looking at options for such a method.

File: src/codegen/impl_debug.rs (hunk 2 above)
Comment (chrysn) [reply src/codegen/impl_debug.rs]: I don't see how, it seems to match the rest of the style to me. The old code gained an indent level of 4 spaces due to the if, and the new code mirrors that.

**Comment by chrysn** [comment]
I played around with how the two detail-hiding parts could be mitigated, but the Array implementation (with the String part) only triggers for large arrays (>32 members), and for the Vector part I could not produce a .h file that would trigger that code path. Unless reviewers insist, I'd leave those as they are, trimming the debug output down when `--use-core` is given.

As for the test suite, all the errors that show are `do catch` related, with me not touching code near anything that looks like that; it appears that it's already broken in master.

That being said, I'd propose this status as a solution to #1100; I'm happy to incorporate any changes that are suggested. (Especially, the concrete placeholder values could be changed; I think "[...]" is fine, but "Array with length x" and "vector" might make it easier to spot the code that's responsible).

File: src/codegen/impl_debug.rs (hunk 2 above)
Comment (emilio) [reply src/codegen/impl_debug.rs]: The parenthesis are four spaces to the right:

```
                        Some((
                            format!("{}: [...]", name), vec![]
                            ))
```

Should be:

```
                        Some((
                            format!("{}: [...]", name), vec![]
                        ))
```

Same below for `Vector`. In any case it's not a huge deal.

**Comment by emilio** [comment]
> As for the test suite, all the errors that show are do catch related, with me not touching code near anything that looks like that; it appears that it's already broken in master.

Yeah, rustfmt is broken on rust nightly, thus the failures. I can merge manually if it's not gone by the time the test is added.

> That being said, I'd propose this status as a solution to #1100; I'm happy to incorporate any changes that are suggested. (Especially, the concrete placeholder values could be changed; I think "[...]" is fine, but "Array with length x" and "vector" might make it easier to spot the code that's responsible).

Yeah, this looks great, it's definitely an improvement over "doesn't build" :P. It should be easy to improve on this should it be needed.

**Comment by chrysn** [comment]
With the new version, the indenting is fixed.

For tests, I've picked derive-debug-bitfield as a staring point and added a derive-debug-bitfield-core, as that makes good visual comparison between the core and std versions.

By the way, I did look for what I'd expect to find as the output of a Vector in the tests (`grep '({})' -r .`), but found nought. Is the coverage that's aimed for large enough that this warrants its own issue?

Thanks for your quick reviews.

**Pull request merged**